apps/meteor/client/lib/voip/VoIPUser.ts

```diff
@@ -54,22 +55,160 @@ export class VoIPUser extends Emitter<VoipEvents> implements OutgoingRequestDele
 
 	mediaStreamRendered?: IMediaStreamRenderer;
 
-	private _callState: CallStates = 'IDLE';
+	private _callState: CallStates = 'INITIAL';
 
 	private _callerInfo: ICallerInfo | undefined;
 
 	private _userState: UserState = UserState.IDLE;
 
+	private _connectionState: ConnectionState = 'INITIAL';
+
 	private _held = false;
 
 	private mode: WorkflowTypes;
 
 	private queueInfo: QueueAggregator;
 
+	// private networkMonitor: NetworkMonitor;
+
+	private connectionRetryCount;
+
+	private stop;
+
+	private networkEmitter: Emitter<SignallinSocketEvents>;
+
+	private offlineNetworkHandler: () => void;
+
+	private onlineNetworkHandler: () => void;
+
+	constructor(private readonly config: VoIPUserConfiguration, mediaRenderer?: IMediaStreamRenderer) {
+		super();
+		this.mediaStreamRendered = mediaRenderer;
+		// this.networkMonitor = new NetworkMonitor(this.config.connectionRetryCount);
+		this.networkEmitter = new Emitter<SignallinSocketEvents>();
+		this.connectionRetryCount = this.config.connectionRetryCount;
+		this.stop = false;
+		this.onlineNetworkHandler = this.onNetworkRestored.bind(this);
+		this.offlineNetworkHandler = this.onNetworkLost.bind(this);
+	}
+
+	/**
+	 * Configures and initializes sip.js UserAgent
+	 * call gets established.
+	 * @remarks
+	 * This class configures transport properties such as websocket url, passed down in config,
+	 * sets up ICE servers,
+	 * SIP UserAgent options such as userName, Password, URI.
+	 * Once initialized, it starts the userAgent.
+	 */
+
+	async init(): Promise<void> {
+		const sipUri = `sip:${this.config.authUserName}@${this.config.sipRegistrarHostnameOrIP}`;
+		const transportOptions = {
+			server: this.config.webSocketURI,
+			connectionTimeout: 100, // Replace this with config
+			keepAliveInterval: 20,
+			// traceSip: true
+		};
+		const sdpFactoryOptions = {
+			iceGatheringTimeout: 10,
+			peerConnectionConfiguration: {
+				iceServers: this.config.iceServers,
+			},
+		};
+		this.userAgentOptions = {
+			delegate: {
+				onInvite: async (invitation: Invitation): Promise<void> => {
+					await this.handleIncomingCall(invitation);
+				},
+			},
+			authorizationPassword: this.config.authPassword,
+			authorizationUsername: this.config.authUserName,
+			uri: UserAgent.makeURI(sipUri),
+			transportOptions,
+			sessionDescriptionHandlerFactoryOptions: sdpFactoryOptions,
+			logConfiguration: false,
+			logLevel: 'error',
+		};
+
+		this.userAgent = new UserAgent(this.userAgentOptions);
+		this.userAgent.transport.isConnected();
+		this._opInProgress = Operation.OP_CONNECT;
+		try {
+			this.registerer = new Registerer(this.userAgent, { expires: 60 });
+			this.userAgent.transport.onConnect = this.onConnected.bind(this);
+			this.userAgent.transport.onDisconnect = this.onDisconnected.bind(this);
+			window.addEventListener('online', this.onlineNetworkHandler);
```

Comment:
There's just one scenario that i think this won't work, and it's when asterisk ip is blocked by a firewall. In that case, the browser will be "online" but asterisk won't be reachable. Not sure how to overcome that tho

Reply:
Connection is going to happen in this.userAgent.start() function. When the connection could not be established for any reason such as firewall blocking the address, ideally userAgent.start() should throw exception. We are catching that exception. I have tested this function when there is no network and it throws an exception. I will also test it for firewall scenario.

Reply:
Yeah, after testing, I think that if a customer has asterisk IP blocked by firewall, then it's beyond our reach. Client will attempt reconnections to asterisk and will fail, but since that will be on customer's infrastructure, we don't have power to decide there.

@amolghode1981 , what do you think on sharing this as part of onboarding to customers? 👀

Reply:
Yeah we should notify the customer about this.